packages/nx/src/lock-file/npm-parser.ts

```diff
@@ -119,6 +109,17 @@ function addNodes(
       }
     });
   }
+
+  // some packages can be both hoisted and nested
+  // so we need to run this check once we have all the nodes and paths
```

Comment:
We are marking hoisted nodes and adding nodes to graph after all of them were collected. Same as it's done in Yarn and Pnpm.